mlflow/pytorch.py

```diff
@@ -0,0 +1,145 @@
+
+
+def _load_model(path, **kwargs):
+    mlflow_model_path = os.path.join(path, "MLmodel")
+    if not os.path.exists(mlflow_model_path):
+        raise RuntimeError("MLmodel is not found at '{}'".format(path))
+
+    mlflow_model = Model.load(mlflow_model_path)
+
+    if FLAVOR_NAME not in mlflow_model.flavors:
+        raise ValueError("Could not find flavor '{}' amongst available flavors {}, "
+                         "unable to load stored model"
+                         .format(FLAVOR_NAME, list(mlflow_model.flavors.keys())))
+
+    # This maybe replaced by a warning and then try/except torch.load
+    flavor = mlflow_model.flavors[FLAVOR_NAME]
+    if torch.__version__ != flavor["pytorch_version"]:
+        raise ValueError("Unfortunately stored model version '{}' does not match "
+                         "installed PyTorch version '{}'"
+                         .format(flavor["pytorch_version"], torch.__version__))
+
+    path = os.path.abspath(path)
+    path = os.path.join(path, mlflow_model.flavors[FLAVOR_NAME]['model_data'])
+    return torch.load(path, **kwargs)
+
+
+def load_model(path, run_id=None, **kwargs):
+    """
+    Load a PyTorch model from a local file (if run_id is None) or a run.
+    :param path: Local filesystem path or Run-relative artifact path to the model saved
+                 by `mlflow.pytorch.log_model`.
+    :param run_id: Run ID. If provided it is combined with path to identify the model.
+    :param kwargs: kwargs to pass to `torch.load` method
+    """
+    if run_id is not None:
+        path = mlflow.tracking._get_model_log_dir(model_name=path, run_id=run_id)
+
+    return _load_model(path, **kwargs)
+
+
+def load_pyfunc(path, **kwargs):
+    """
+    Load the model as PyFunc. The loaded PyFunc exposes a `predict(pd.DataFrame) -> pd.DataFrame`
+    method that, given an input DataFrame of n rows and k float-valued columns, feeds a
+    corresponding (n x k) torch.FloatTensor (or torch.cuda.FloatTensor) as input to the PyTorch model. 
+    ``predict`` returns the model's predictions (output tensor) in a single-column DataFrame.
+
+    :param path: Local filesystem path to the model saved by `mlflow.pytorch.log_model`.
+    :param kwargs: kwargs to pass to `torch.load` method.
+    """
+    return _PyTorchWrapper(_load_model(os.path.dirname(path), **kwargs))
+
+
+class _PyTorchWrapper(object):
+    """
+    Wrapper class that creates a predict function such that
+    predict(data: pd.DataFrame) -> model's output as pd.DataFrame (pandas DataFrame)
+    """
+    def __init__(self, pytorch_model):
+        self.pytorch_model = pytorch_model
+
+    def predict(self, data, device='cpu'):
+        if not isinstance(data, pd.DataFrame):
+            raise TypeError("Input data should be pandas.DataFrame")
+        self.pytorch_model.to(device)
+        self.pytorch_model.eval()
+        with torch.no_grad():
+            input_tensor = torch.from_numpy(data.values.astype(np.float32)).to(device)
```

Comment:
Sorry I'm noticing this late, wanted to get your opinion: currently users can only predict on models with a single input tensor, which precludes using MLflow with models that take input tensors of different types (e.g. when predicting on images, maybe we'd have a float tensor containing pixel values & integer values corresponding to image height/width). IMO this restriction is fine as long as we can extend this implementation to support multi-input models in the future.

My only concern with the current implementation is that we convert the entire DataFrame to a single tensor & pass that as the sole input to the model, which requires us to make breaking changes down the line if we want to instead pass each column as an input tensor to the model. Maybe we should instead require that the input DataFrame `df` contains a single column of numpy arrays & convert that column to an input tensor (e.g. via `np.hstack(data[data.columns[0]])`?

Reply:
Discussed offline: let's keep this issue in mind but move forward with the current implementation for now :)